nighthawk/test/stream_decoder_test.cc

```diff
@@ -0,0 +1,136 @@
+
+  Envoy::Thread::ThreadFactoryImplPosix thread_factory_;
+  Envoy::Event::RealTimeSystem time_system_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  Envoy::Api::Impl api_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  StreamingStatistic connect_statistic_;
+  StreamingStatistic latency_statistic_;
+  Envoy::Http::HeaderMapImpl request_headers_;
+  uint64_t stream_decoder_completion_callbacks_;
+  uint64_t pool_failures_;
+};
+
+TEST_F(StreamDecoderTest, HeaderOnlyTest) {
+  bool is_complete = false;
+  auto decoder =
+      new StreamDecoder(*dispatcher_, time_system_, *this, [&is_complete]() { is_complete = true; },
+                        connect_statistic_, latency_statistic_, request_headers_, false);
+  auto headers = std::make_unique<Envoy::Http::HeaderMapImpl>();
+  decoder->decodeHeaders(std::move(headers), true);
+  EXPECT_TRUE(is_complete);
+  EXPECT_EQ(1, stream_decoder_completion_callbacks_);
+}
+
+TEST_F(StreamDecoderTest, HeaderWithBodyTest) {
+  bool is_complete = false;
+  auto decoder =
+      new StreamDecoder(*dispatcher_, time_system_, *this, [&is_complete]() { is_complete = true; },
+                        connect_statistic_, latency_statistic_, request_headers_, false);
+  auto headers = std::make_unique<Envoy::Http::HeaderMapImpl>();
+  decoder->decodeHeaders(std::move(headers), false);
+  EXPECT_FALSE(is_complete);
+  Envoy::Buffer::OwnedImpl buf(std::string(1, 'a'));
+  decoder->decodeData(buf, false);
+  EXPECT_FALSE(is_complete);
+  decoder->decodeData(buf, true);
+  EXPECT_TRUE(is_complete);
+  EXPECT_EQ(1, stream_decoder_completion_callbacks_);
+}
+
+TEST_F(StreamDecoderTest, TrailerTest) {
+  bool is_complete = false;
+  auto decoder =
+      new StreamDecoder(*dispatcher_, time_system_, *this, [&is_complete]() { is_complete = true; },
+                        connect_statistic_, latency_statistic_, request_headers_, false);
+  auto trailers = std::make_unique<Envoy::Http::HeaderMapImpl>();
+  decoder->decodeTrailers(std::move(trailers));
+  EXPECT_TRUE(is_complete);
+  EXPECT_EQ(1, stream_decoder_completion_callbacks_);
+}
+
+TEST_F(StreamDecoderTest, LatencyIsNotMeasured) {
+  auto decoder = new StreamDecoder(*dispatcher_, time_system_, *this, []() {}, connect_statistic_,
+                                   latency_statistic_, request_headers_, false);
+  Envoy::Http::MockStreamEncoder stream_encoder;
+  Envoy::Upstream::HostDescriptionConstSharedPtr ptr;
+  EXPECT_CALL(stream_encoder, encodeHeaders(HeaderMapEqualRef(&request_headers_), true));
+  decoder->onPoolReady(stream_encoder, ptr);
+  auto headers = std::make_unique<Envoy::Http::HeaderMapImpl>();
+  decoder->decodeHeaders(std::move(headers), true);
+  EXPECT_EQ(0, connect_statistic_.count());
+  EXPECT_EQ(0, latency_statistic_.count());
+}
+
+TEST_F(StreamDecoderTest, LatencyIsMeasured) {
+  auto decoder = new StreamDecoder(*dispatcher_, time_system_, *this, []() {}, connect_statistic_,
```

Comment:
Who is going to be the ultimate owner of `StreamDecoder`? I'm curious if we can do a more robust ownership management here with smart pointers, generally we shy away from raw pointer in Envoy, but I agree that sometimes it makes sense. Just want to see if we can avoid potential leak scenarios.

Reply:
I was considering adding a `StreamDecoderPool`; this may help avoid some allocations in the fast path, and would also be a natural place to take ownership. wdyt?

Reply:
Sure, that could be something to consider as an optimization. I'm curious about what to do even before we have this, is it possible that the allocating entity can also do cleanup via a smartptr, or do we need to self-destruct?